부산대 BE_이시웅 2단계 - 회원 로그인

silverttthin opened the pull request:
최소기능들만을 구현해 너무 보잘것 없어 보여 걱정됩니다.
일단 제출 데드라인을 지키기 위해 최대한 18시까지 제출을 위해 가장 필수로 요구되는 최소 기능들인 회원가입, 로그인(by jwt)으로 구현했습니다.
멘토님께서 리뷰를 해주시기까지의 텀 동안 추가적으로 커밋해볼 계획이며 피드백도 빡세게 반영하겠습니다. 
개선사항, 잘못된 부분 등을 빡세게 리뷰해주시면 감사할 것 같습니다.

일단 pr 직후 제가 진행해 나갈 예상 테스크들은 다음과 같습니다.
- 게시글에 작성자 fk 넣어 게시자 기능 도입해 헤더에 들어있는 토큰을 읽고 수정,삭제 api에 검증 로직 도입
- e2e 테스트에 유저 플로우까지 도입

7/8 03:17) 인터셉터를 이용해 컨트롤러 인자에서 로그인한 유저의 pk를 얻는 단계까지 수행했습니다. 다음날 곧바로 기존 아이템 서비스 로직에 인가로직또한 추가하겠습니다!

16:25) 인가로직 및 자잘한 기능 추가하였고 최종적으로 유저가 반영된 e2e 테스트까지 작성했습니다. 개인적인 추가구현은 멈추고 피드백을 반영하겠습니다

사소한 궁금한 점이 4가지 생겼습니다
1. 전 단계 코멘트 달아주셔서 User엔티티에 인자가 다른 생성자를 2개를 도입하고 userId 외래키도 유저 엔티티에 추가했습니다. 생성자 2개 + 객체의 속성이 많아져 positional argument 특성 때문에 불편함도 많이 느꼈습니다. 이를 어떻게 개선해야 할까요? (Lombok의 @Builder를 도입하면 positional argument 특성에 의한 불편함은 확실히 사라질 것 같은데 생성자가 여러개인 경우에도 효과가 있는진 잘 모르겠습니다.)

2. 이전도 그렇고 이번도 그렇고 기능을 다 구현하고 마지막에 e2e 테스트코드를 최종 커밋으로 작성했습니다. 순서가 잘못되어서 제가 테스트 코드를 작성하는 효용을 못느끼는 건지 궁금합니다. 알고리즘 문제 풀때처럼 타이핑 전에 계획을 짜는게 정답일까요? 즉 어떻게 코드 작성하고 어떤 기능 명세를 낼건지를 먼저 계획을 짜고 그 다음에 e2e 테스트를 작성하고 본격적인 기능구현으로 들어가는게 진짜 테스트 코드 작성을 공부하는 왕도인가요?

3. 필터와 인터셉터 결국 컨트롤러 이전에서 보안과 같은 공통 관심사를 처리하는 로직을 작성하는 레이어로 학습했습니다. 다만 필터는 인터셉터보다 더 앞에 있다는 것, 그리고 스프링 기술이 아닌 톰켓 기술이라는 점(물론 최신문법으론 스프링 문법으로 다룰 수 있다곤 하지만)을 학습했습니다. 이 컨트롤러 앞단에서 어떤 작업을 처리할 때 둘 중 하나를 선택하는 원칙, 선택기준같은게 있을까요?

4. 스프링 시큐리티를 제대로 공부하거나 익히고 쓴적이 없어서 잘 모르겠습니다. 간단하고 미니멀한 인증/인가 기능이지만 단방향 암호화(물론 crypto 라이브러리도 시큐리티 부분집합이지만..), jjwt 라이브러리를 이용한 토큰 기능, 인터셉터를 이용한 컨트롤러 인증 적용 등을 스프링 시큐리티 종속성 없이도 구현이 가능하다는 판단이 확립되게 되는 step2인 것 같습니다. 그럼에도!!!!! 시큐리티를 무조건 쓰는 키포인트가 있을 것이고 그러기에 시큐리티가 중요하고 어렵고 많이 쓰인다고 생각합니다. 시큐리티에서 제공하는 가장 핵심적이고 유용한 기능들은 어떤 점들이 있나요?

silverttthin committed:
feat: Item의 allArg 생성자 팩토리 메서드화
- 검증로직을 안에 삽입
- pk없는 생성자도 이 팩토리 메서드를 호출하는 것으로 변경
- 추가로 인텔리제이의 코드 컨벤션 기능 사용해 개행 조절했음

silverttthin committed:
docs: 요구사항 추가

silverttthin committed:
feat: 유저 엔티티 추가 및 디비 초기화 반영

silverttthin committed:
feat: 유저 도메인 레이어드 구조 작성 및 회원가입 기능 구현

silverttthin committed:
build: jwt 의존성 설치 및 환경변수 작성
- properties 문서는 멘토님 테스트 용이성 목적도 있고 해서 gitignore에 아직 넣진 않았습니다.

silverttthin committed:
feat: UserRepository 코드 작성
- 어차피 에러 컨트롤을 optional로 다 하므로 잘못되면 모두 try-catch를 통해 null 리턴하게 세팅

silverttthin committed:
feat: jwt 유틸객체 작성
- 토큰 메서드 작성
- 토큰값과 만료시간은 환경변수에서 가져오기(gitignore 추가는 아직)
- Jwts의 deprecated 기능을 대체해 최신 문법으로 빌더 체이닝 작성하였음

silverttthin committed:
feat: 로그인 기능 구현

silverttthin committed:
fix: schema 및 초기화 로직 변경

silverttthin committed:
fix: 기존 로그인에서 유저를 가져오던 기능을 email 칼럼만 사용으로 변경 + 칼럼명 틀려서 발생했던 버그 원인 수정(gpt를 쓸거면 제대로 씁시다..!)
- 바꾼 이유는 email로 찾아온 객체의 password 속성을 req.password()와 비교해도 로직은 동일하고 아주 작겠지만 쿼리 속도가 조금 더 빠를 것 같아서 변경했습니다.

silverttthin committed:
feat: 로그인/회원가입에 비밀번호 단방향 암호화 기능 구현
- 처음엔 MessageDigest 내장 클래스로 직접 암호화를 할까 싶었습니다.
- 다만 보안 관련은 어설픈 직접 구현보단 이미 최대한 잘 만들어진 것들을 가져다 쓰는게 정답이기도 하고 crypto 의존성은 간단한 인코더만으로 salt까지 잘 적용돼있어 의존성을 설치해 사용하는 방향으로 결정했습니다!

silverttthin committed:
chore: 디렉토리 명 변경.. 더 깔끔하고 나은 패키지 정리를 어떻게 해야할까.. config, util, 그리고 앞으로 추가될 다양한 도메인들..

silverttthin committed:
refactor: 중복된 이메일/닉네임은 단순 400 대신 conflict를 의미하는 409 리턴하도록 구현
- DataIntegrityViolationException은 무결성 조건 위배 시 쓰는 내장예외
(이메일과 닉네임은 unique)

silverttthin committed:
feat: 토큰 파싱 유틸함수 작성
- 토큰을 파싱하지만 리턴값을 사용하지 않고 그대로 리턴하는게 킥임
- 왜냐하면 중간에 parse()가 예외를 터트린다 == 잘못된 토큰
- 무사히 다 끝난다 -> 유효한 토큰임을 true 리턴으로 표시

silverttthin committed:
chore: 유저 목업데이터 password를 입력값 1을 해싱한 값으로 변경

silverttthin committed:
feat: jwt 토큰의 sub을 user pk로 변경 + 파싱해 pk 얻는 유틸함수 구현

silverttthin committed:
feat: 인터셉터 구현 및 등록
- 일단 로그인,회원가입 제외 모든 api는 로그인 유저만 사용하게 세팅
- 게시글 조회는 뺴고 싶었는데 조회와 생성이 메소드만 다르고 경로가 같아서 경로 기준 화이트리스트로 구분하는건 불가능해 일단 미룸

mkom76 commented on src/main/java/gift/product/dto/CreateUserRequest.java:
```diff
@@ -0,0 +1,18 @@
+
+
+public record CreateUserRequest(
+	@Email(message = "올바르지 않는 이메일 형식입니다.")
```
바로 유효성 검사 적용해주셨네요👍👍

mkom76 commented on src/main/java/gift/product/entity/Item.java:
```diff
@@ -11,19 +11,19 @@ public class Item {
+		this(null, name, price, imageUrl);
+	}
+
 	public static void validateKakaoKeyword(String name){
```
이 유효성 로직은 이제 내부에서만 쓰기 때문에 private으로 선언해도 될 것 같아요~!

mkom76 commented on src/main/java/gift/product/etcs/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,52 @@
+
+	@Override
+	public void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler, ModelAndView modelAndView) throws Exception {
+		HandlerInterceptor.super.postHandle(request, response, handler, modelAndView);
```
이 postHandle과 afterCompletion이 어떤 동작을 하는지 설명해주실 수 있을까요?😊

silverttthin replied:
인터셉터는 HandlerInterceptor 인터페이스를 구현합니다. 인터셉터는 컨트롤러 앞단에서 요청과 응답에 대해 처리하는 레이어이며 인터페이스에서는 3가지 메서드를 구현하라 지시돼있습니다.
1. preHandle: 우리가 쓰는 함수로 컨트롤러 실행 직전, 즉 컨트롤러 작업전에 처리해야할 인증인가 등을 처리할 때 써야하는 함수입니다
2. postHandle: mvc 패턴에서 모델 조립되고 뷰가 조립되기 직전 타이밍의 작업을 정의합니다.
3. afterCompletion: 이름대로 다 끝난, 즉 뷰 렌더링 직후 타이밍의 작업을 정의합니다.

mkom76 commented on src/main/java/gift/product/etcs/JwtUtil.java:
```diff
@@ -0,0 +1,88 @@
+	private final SecretKey secretKey;
+	private final long expirationMs;
+
+	public JwtUtil(@Value("${jwt.secret}") String secretKey, @Value("${jwt.expiration}") long expirationMs) {
```
secret을 properties에서 관리👍👍 Value 어노테이션을 잘 활용해보셨네요.

mkom76 commented on src/main/java/gift/product/etcs/JwtUtil.java:
```diff
@@ -0,0 +1,88 @@
+		Date now = new Date();
+		Date expiryDate = new Date(now.getTime() + expirationMs);
+
+		return Jwts.builder()
```
빌더 패턴을 잘 사용하셨네요! 자바 코드에 많이 등장하니 잘 익혀보시면 좋겠습니다😊

mkom76 commented on src/main/java/gift/product/service/UserService.java:
```diff
@@ -0,0 +1,60 @@
+		if(userRepository.findByNickname(req.nickName()).isPresent())
+			throw new DataIntegrityViolationException("이미 사용중인 닉네임입니다.");
+
+		String encodedPassword = passwordEncoder.encode(req.password());
```
패스워드 암호화 후 저장👍👍

mkom76 commented on src/main/java/gift/product/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,102 @@
+	public Long save(User user) {
+		final String sql = "INSERT INTO users (email, password, nickname) VALUES (?, ?, ?)";
+
+		KeyHolder keyHolder = new GeneratedKeyHolder();
```
KeyHolder를 잘 사용하고 계시네요!😊 이거의 동작방식과 왜 사용하는지를 공유해주실 수 있을까요~?

silverttthin replied:
jdbc에서 auto generated된 key값을 가져올 때 사용합니다. 제 코드에선 save()에서 생성된 pk를 리턴하기 위해 사용합니다.

여기까진 알고, 즉 왜 쓰는진 인지했는데 동작방식은 몰라서 조금 검색해보고 다음처럼 요약해봤습니다.
- 일반적으로 db들은 insert 동작 후 auto generated key를 리턴
- jdbc 드라이버는 insert query 실행 후 리턴된 이 pk값을 받아놓는 구현체인 KeyHolder에 저장

silverttthin replied:
각 디비 드라이버별로 구현이 다르고 지원도 다르다고 하기에 h2 깃허브 오픈소스 들어가서 구현을 살펴보려 했는데.. 아직 코드보는 짬밥이 적어서인지 JdbcPreparedStatement.java 나가떨어졌습니다.. 
핵심은 키값을 가져와 키홀더에 넣는 책임주체는 jdbc driver란 것을 확인했습니다

mkom76 commented on src/main/java/gift/product/etcs/WebConfig.java:
```diff
@@ -0,0 +1,24 @@
+	public void addInterceptors(InterceptorRegistry registry) {
+		registry.addInterceptor(authenticationInterceptor)
+			.addPathPatterns("/**")
+			.excludePathPatterns("/users/**");
```
회원가입, 로그인 제외한 api에 토큰을 요구👍

silverttthin committed:
feat: item 엔티티에 작성자 속성 구현

silverttthin committed:
refactor/test: 작성자 반영된 리포지토리 코드 수정 및 테스트 작상

silverttthin committed:
feat: 상품 기능에 작성자 검증로직 추가

silverttthin committed:
chore: 임시 주석처리

silverttthin committed:
feat: 인증 우회를 어노테이션으로 구현
- webconfig에 등록 시 경로 기준으로 인터셉터를 등록하는 것이기에 같은 경로에 메서드만 다른 게시글 생성/조회를 구분하기가 불가능해서 임시로 조회도 인증을 요구하게 했음
- 인터셉터 내부 로직에서 해당 어노테이션이 붙은 핸들러일 때 토큰 작업을 수행하는 것으로 이를 우회해 구현했음

silverttthin committed:
refactor: yml구조로 변경

silverttthin committed:
fix: 잘못된 생성자 사용!!!!!!!
- 왜 빌더를 쓰는지 step2 하면서 너무 체감된다

silverttthin committed:
test: 유저 적용된 e2e 테스트 작성

mkom76 commented on src/main/java/gift/product/entity/Item.java:
```diff
@@ -3,25 +3,28 @@
-	private Long id;
+	private final Long id;
+
+	private final Long userId;
```
보통 도메인의 생성자를 나타낼 때 String createdBy;를 주로 사용합니다. 
User 도메인과 외래키로 종속성을 주려는 목적으로 보이는데요, 현업에서는 테이블 간 종속성을 최소화하기 위해(시스템이 커질수록 종속성이 있는 경우 많은 제약이 생겨요. 현업에서 RDB가 많이 사라진 이유와 맞닿습니다) String으로만 저장하고 User에서 검색하는 방식을 사용합니다. 참고해주세요~

silverttthin replied:
제가 이전에 진행한 프로젝트에선 각 도메인 테이블간 외래키 연결을 다 지정하고 jpa 양방향까지 해뒀는데 꽤 고생을 했던 기억이 있습니다. 엄청 불편했는데도 그게 정석일것 같았고 그냥 jpa의 기능을 온전히 활용한다고 생각했어서 잘못했는지 마냥 확신이 들진 않았는데 이제 좀 생각이 확립이 됩니다..! 그냥 외래키를 string으로 지정하고 서비스에서 쿼리날려서 가져오는 방식으로 하는군요. 앞으로 플젝을 할 땐 테이블간 종속성을 확 줄여야겠습니다.

그렇다면 역으로 궁금한 게 '그럼에도 연관관계를 맺어야 하는 경우'는 어떤 경우인가요? 
제 생각엔.. 이를테면 두 묶음의 속성들의 차이가 어느정도 존재해서(?) 게시판 엔티티를 두개의 테이블로 분류해서 속성들을 저장한다고 가정한다면, 두개의 게시판 테이블은 서비스 코드에서 빈번하게 같이 활용될 것이므로 이 경우엔 외래키로 묶어둬야할 것 같습니다.

+) 현업에선 그럼 rdb 대신 몽고디비같은 nosql을 주로 쓰나요?

mkom76 replied:
조금만 더 부연설명드리면, 제1정규화, 제2정규화 등 학교에서 배운 내용들을 잘 살펴보면 데이터 정합성과 리소스를 절감하는데 큰 목표가 있습니다. 그런데 현실에서는 리소스 효율이 많이 올라가서, 데이터를 여러 군데에 중복해서 저장하고 여러 형태로 저장할 수 있는 여건이 되었어요.
말씀대로 현업에서는 Nosql도 많이 사용하고요, RDB를 의존성을 최소화하여 사용하기도 합니다.
비즈니스상 연관관계가 큰 테이블, 혹은 테이블 하나가 데이터가 커져서 두개로 나누는 경우 등에 의존성을 넣기는 합니다. 사실 DB 설계도 완벽한 정답은 없습니다. 비즈니스가 어떻게 튈지 모르기 때문이에요. 다만 최근에는 의존성을 많이 줄이고 있다 정도로 이해를 하시면 좋을 것 같아요😊😊

mkom76 commented on src/main/java/gift/product/service/ItemService.java:
```diff
@@ -18,45 +21,49 @@
-		itemRepository.findById(itemId).orElseThrow(() -> new RuntimeException("존재하지 않는 아이템입니다."));
+	public GetItemResponse updateItem(Long itemId, Long userId, ItemRequest req) {
+		Item item = itemRepository.findById(itemId).orElseThrow(() -> new RuntimeException("존재하지 않는 아이템입니다."));
+		if(item.getUserId() != userId)
```
Id 검증 좋습니다!👍
아래와 중복되니 메소드로 빼도 좋겠네요. 그리고 if 문에 중괄호는 한줄이어도 붙이는게 컨벤션입니다~ 참고해주세요!

silverttthin replied:
컨벤션이였다니.. 곧바로 제 스타일을 바꿔야겠습니다! 원랜 코드라인 개행 하나라도 아득바득 줄이려는 오랜 버릇이였습니다

silverttthin replied:
```
private void isItemAuthor(Long itemAuthorId, Long userId) {
		if(itemAuthorId != userId){
			throw new IllegalArgumentException("작성자만 수정,삭제 가능합니다.");
		}
	}
```

추가로 다음처럼 중복되는 로직을 메소드로 빼 작성자 == 로긴유저 검증을 리팩토링했습니다.
근데 이 메소드도 itemService에 있는게 맞는지 궁금합니다. 제 머릿속에 다음처럼 2가지 생각이 공존합니다
**작성자==로긴유저 검증도 엄연히 서비스 로직이니 있는게 맞다** vs **메인 서비스 로직이 아니므로 다른 곳에 작성하는게 맞다. (근데 그럼어디..? validator?)**

일단 제 느낌은 전자쪽에 가깝습니다만.. 사바사 부바부 케이스 바이 케이스 일것 같은 느낌입니다

mkom76 commented on src/main/java/gift/product/controller/ItemController.java:
```diff
@@ -21,6 +22,7 @@ public class ItemController {
 
 
 	// 게시글 생성
+	@Authenticated
```
흔히 겪는 문제를 잘 우회해주셨네요!👍
4번의 답변이 될 것 같은데요,
이러한 복잡한 권한 관련 문제의 솔루션들을 spring security에서 많이 해결해줍니다. 메소드별로 제어도 가능하고, user의 role에 따라 제어도 가능합니다. 현업에서도 이 기능을 많이 사용해요😊
코드를 많이 줄여주고 추상화되어 개발자가 인증인가에 너무 많은 시간을 사용하지 않도록 도와줍니다. 참고해주세요~

mkom76 commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,16 @@
+spring:
```
yaml 형식은 나중에 k8s를 공부할때도 유용하니 잘 익혀두세요~👍 좋은 경험이 되셨을 것 같습니다!

mkom76 commented on src/test/java/gift/E2ETest.java:
```diff
@@ -33,79 +37,108 @@ void setUp() {
 	@Test
-	void 상품CRUD전체플로우() {
-		ItemRequest request = new ItemRequest("콜라", 1500, "url");
+	@DisplayName("인증인가 적용된 상품 e2e 테스트")
```
E2E 테스트까지 잘 작성해주셨네요! 참고로 실제 테스트용 서버에서는 DB에 테스트용 유저를 만들어두고 테스트 시 그 유저로 토큰을 획득하여 E2E 테스트를 수행합니다. 참고해주세요~!

mkom76 reviewed (changes requested):
역시나 많이 고민하시면서 많은 기능들을 경험해보려는 노력이 많이 보입니다😊😊 고생하셨어요.
코멘트 몇개 드렸는데 확인 부탁드려요~
그리고 etcs 패키지는 config, common 패키지로 나누어도 될 것 같아요! etcs는 잘 사용하지 않는 패키지명입니다😊
질문의 대한 답변도 드릴게요.
1. 필드 수가 많아지면 말씀대로 빌더 패턴이 굉장히 효과적입니다! 빌더 패턴에는 생성자가 여러개일 필요가 없어질 것 같은데 제가 질문을 잘못 이해한건지 모르겠어요~ 말씀나온김에 한번 실제로 구현해보시면서 저에게 코멘트 부탁드려요!
2. 테스트를 먼저 짜고 코드를 짜는게 왕도라기보다는 여러 기법 중 하나입니다~ 나중에 TDD라는 기법을 배우시면 테스트 코드를 먼저 짜는게 어떤 이점이 있는지 알게 되실거에요! 꽤 오래전부터 각광받았던 기법이긴 합니다. 그리고 현업에서 일반적으로 많이 느끼는 테스트의 효용은 커밋 전 테스트를 돌려보는겁니다ㅎㅎ 에러가 발생하면 바로 처리하기 쉽도록요!
3. 필터는 일반적으로 요청 데이터 자체에 대한 핸들링을 진행하고요(인코딩, 데이터압축, 데이터 형식 변화 등), 인터셉터는 어플리케이션과 관련된 핸들링을 합니다(예외발생 처리, 스프링 빈 직접 사용 등) 다만 보안/인증/인가는 두 곳 모두에서 진행할 수 있어요😊 참고해주세요!
4. 이건 코멘트 참고해주세요~

silverttthin committed:
chore: 디렉토리 구조 개편

silverttthin committed:
fix: 엔티티 내에서만 쓰는 메서드 접근지시자를 private로 변경

silverttthin committed:
chore: 중괄호 컨벤션 수정

silverttthin committed:
docs: 누락했던 readme 요구사항 체크 완료

silverttthin committed:
refactor: 아이템 게시자가 현재 로그인한 유저인지 검증하는 작업을 private 메소드로 묶어 사용

silverttthin committed:
refactor: 롬복 설치해 생성자를 빌더 패턴으로 개편

silverttthin commented:
피드백 다 반영하고 생성자->빌더패턴까지 적용했습니다! 확실히 불편함을 느끼고 롬복을 적용하니 체감이 크게 느껴집니다. 다만 카테캠은 학습의 과정이다보니 getter, 생성자 어노테이션은 사용하지 않으려합니다.

멘토님께서 생성자 여러개일 필요가 없다고 생각한 부분에 대해선 제가 질문을 할때 컨텍스트를 제대로 전달치 못한 탓입니다 ㅜ 저 질문의 배경은 다음과 같았습니다.

1. Builder 어노테이션은 엔티티 클래스 자체에 걸든가 아니면 생성자를 직접 작성해 그 메서드에 걸게 할 수 있습니다. 
2. 후자를 택한 경우 빌더패턴으로 접근 가능한 속성은 그 생성자의 패러미터입니다.
3. 저는 생성자를 사용 시 (대부분이 그러하듯 pk의 생성은 dbms에 전적으로 맡기는 환경에서) pk를 개발자가 직접 작성할 일이 없을 것이고 오히려 그럼 안된다 생각해 이전에 한 개인 플젝에선 모든 엔티티의 빌더 생성자에 ``Long id``를 배제했습니다.
4. 다만 이전 단계에서 pk를 직접 넣은 생성자를 사용해 객체를 생성해야하는 과정이 있었습니다. (``itemService.updateItem()``에서 updatedItem 객체를 생성하는 부분입니다. 지금은 빌더로 대체)
5. 그래서 전 첫 질문 때 생성자가 지금 2개이니 2개다 builder 걸어야 하고 그럼 생성자가 여러개라 헷갈리는 근본적인 문제는 해결되지 않을텐데? 라는 생각에서 비롯된 질문이였습니다

지금은 개발의 용이성을 위해 그냥 Item 클래스 팩토리 메서드 하나만 남기고 거기에 builder을 걸었습니다. 이 모든 문제의 원인은 제가 갖고있던 생각(위의 3번)이였습니다. 멘토님께서 이에 대해 어떻게 생각하시는지 궁금합니다!!

mkom76 commented on src/main/java/gift/product/entity/Item.java:
```diff
@@ -27,7 +27,7 @@ public Item(Long userId, String name, Integer price, String imageUrl) {
 	}
 
-	public static void validateKakaoKeyword(String name){
+	private static void validateKakaoKeyword(String name){
```
Static도 빼도 될 것 같아요~

mkom76 commented on src/main/java/gift/product/service/ItemService.java:
```diff
@@ -62,11 +60,14 @@ public GetItemResponse updateItem(Long itemId, Long userId, ItemRequest req) {
 		itemRepository.deleteById(itemId);
+	}
 
+	private void isItemAuthor(Long itemAuthorId, Long userId) {
```
중복코드를 잘 없애주셨네요!👍
코멘트에 남기신 질문에 대한 답을 여기서 드리자면, 서비스레이어에 두어도 괜찮기는 합니다. 다만 좀더 객체지향적으로 작성한다고 하면, Item 클래스 내부에서 확인 후 에러를 던지도록 하는것도 방법이에요. 메소드를 만들고 item.메소드() 로 서비스에서 호출만 하는거죠. Item의 수정/삭제에 대한 유효성 검사 책임을 Item이 가지도록 하는 방법입니다😊

silverttthin replied:
이 또한 `validateKakaoKeyword(name)`처럼 아이템 도메인과 관련된 동작을 아이템 엔티티 클래스 자체에서 가지는 패턴이로군요. 이렇게 도메인관련 동작을 도메인 객체 내에 명시하는게 DDD의 일종인건가요?

silverttthin committed:
feat: 아이템 응답 dto에 작성자 필드 추가
- e2e에도 반영 + 테스트에서 회원가입 플로우 배제하고 테스트 유저 로그인으로 대체
- record에 빌더 붙여 빌더 체이닝으로 생성자 대체

mkom76 reviewed (changes requested):
리뷰 반영하시느라 고생하셨습니다.👍 계속 프로젝트 고도화를 위해 노력하시는 모습이 보이네요😊💯

질문에 대한 답변을 드리자면, requestDto에 id 필드를 두어야 하나? 에 대한 고민을 해보시는 것도 좋겠어요. repositry에서 update하는 sql문을 보면, id를 굳이 dto에 넣을 필요는 없다는 걸 아실 수 있습니다. item.getId()가 아닌, api path variable로 들어온 id를 그대로 넘겨주어서 사용할 수 있겠죠?😊
참고로, 다음 과제에서 JPA를 배울 때도 마찬가지로 update에서는 id를 넘겨줄 필요가 없습니다. 한번 수정해보실까요?😊
말씀하신대로 id는 개발자가 컨트롤할 필요는 없습니다. 클라이언트는 더더욱 알 필요가 없고요. 좋은 고민 포인트입니다😊😊

그리고 builder 패턴은 사실 롬복이 아니라 직접 구현해보시라는 뜻이었는데, 이미 알고 계신 것 같기는 합니다. 가볍게 코멘트로 어떻게 내부 코드가 되어 있는지 달아주세요~😊😊

한번만 더 RC 드리겠습니다!💪💪

silverttthin committed:
chore: static 삭제

silverttthin committed:
refactor: 작성자 검증을 도메인 메서드로 수행하는 것으로 변경

mkom76 commented:
@silverttthin 
네 DTO는 제가 잘못 보았네요.😭 reponseDto에 id가 있고 클라이언트에게 id 정보를 보여주는 설계네요. 혼란 드려서 죄송합니다.
빌더 패턴을 적용하려면, 전체 아규먼트를 받는 생성자를 그대로 두고, id를 제외한 생성자를 만든 후, 그 생성자에 빌더 패턴을 붙히는 건 어떨까요?
<img width="865" alt="image" src="https://github.com/user-attachments/assets/d25ce646-9538-4538-bd5f-711fcdda4578" />

이런 형태입니다ㅎㅎ 이렇게 하면 생성자가 두개여도 조금 명확해지지 않을까 싶은데 혹시 시웅님 의견은 어떠실까요?😊 편하게 의견 주세요~!

silverttthin committed:
refactor: 생성자를 2개로 원상복구 하고 개발자가 쓸 'pk없는 생성자'에만 빌더 지정

silverttthin committed:
fix: 변동된 생성자에 알맞게 수정 메서드의 인자형태 수정 + e2e테스트에도 변화 반영

silverttthin committed:
fix: 서비스 레이어에도 바뀐 수정함수 명세 적용

silverttthin committed:
build: 롬복 삭제

silverttthin committed:
feat: @Builder 쓰던 Item과 아이템 응답 dto에 직접 빌더 구현

silverttthin commented:
@mkom76 
괜찮은 것 같습니다! pk를 인자로 받지않는 생성자만 빌더로 명시해두고 this(...) 생성자로 내부에서 팩토리 생성자 메서드로 넘기면 생성자가 2개여도 빌더로 쓰는건 1개이니 빌더 체이닝을 쓰는 개발자 입장에선 헷갈릴 여지가 없어질 것 같습니다

롬복 빌더만 사용해온 저인지라.. 제 세상에 빌더는 롬복의 @Builder 뿐이였고 그러기에 구현은 몰랐습니다. 그래서 이번 피드백을 기회로 롬복을 삭제하고 직접 빌더를 사용하는 곳에서 빌더 체이닝을 구현해서 커밋했습니다.

검색해보고 구현해보니 난이도는 낮았으나 주요 핵심 2가지가 중요하게 정리되는 것 같습니다.
1. ItemBuilder 객체에서 각 필드별 메서드를 정의해 인자값을 필드로 저장하고(`this.name = name`) **this, 즉 자신 인스턴스를 리턴하는 것이 체이닝 구현의 핵심!!**
2. 다 끝내고 만들어지는건 Item 객체가 아니라 ItemBuilder 객체이다. 이 객체의 데이터를 Item의 생성자를 호출해 넣어서 item 인스턴스를 만드는게 빌더의 마무리 메서드인 `build()`

silverttthin committed:
docs: step3 위시리스트 기능 요구사항 작성

mkom76 reviewed (approved):
리뷰 반영하시느라 고생하셨어요~!👍👍
프로젝트 경험이 이미 있으셔서, 최대한 고도화하려는 노력이 보이네요💯
고생 많으셨어요~! 3단계도 화이팅입니다!💪💪
